Add device_stats table to store device related data for alerting

eskyuu opened the pull request:
This adds a new device_stats table to store device related data for alerting.  Initial data is packet loss and round trip times.

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

eskyuu committed:
Add packet loss column to the devices table

murrant commented:
Ok, so this is a partial fix for something we've needed since the removal of device_perf table.
Another feature that is missing is the traceroute when down. (which should probably be stored in the same place.
We could store it in the device table, or we could store it in a separate table with other stats columns.

FYI, the reason the device_perf table was removed was because it was storing data over time and a lot of it and it needed to be deleted from there from daily.sh otherwise the table blew up.  Also, some of that data was duplicate to what was stored in rrd.

I'm happy to merge this as is, but since there is more to the story and a new table might make sense, I wanted to bring it up.

laf commented:
If we merge as is and change it later it's just a bit disruptive for people as it could be used in alert rules and templates (and device groups). If we moved this to its own table we should then move the other last_ping_ fields as well.

traceroute info making a come back would be good.

eskyuu commented:
Do we want something like a generic device_perf table for all stats?  Or a device_pingstats table that is updated in FpingResponse.php, and then more tables for other data.  

I'm also assuming a 1:1 relationship between the stats table and the device table, and am curious what the advantages are of having more tables vs more columns on the one table.

murrant commented:
Main advantage of having to tables is less data fetched when we fetch with all columns (which is frequently). And saving potential lock contention on writes.  I'm not sure there are enough fields to justify an additional table, just need to think things through.

slashdoom commented:
In our env, alerting-wise, it would help to at least have this be metric-y enough to store _prev values.

ping_rtt_current
ping_rtt_prev
ping_loss_current
ping_loss_prev

So maybe that's closer to a new table?

murrant commented:
Honestly, I kinda hate duplicating rrd data in the database and adding _prev because it is one dimensional...  The real fix would be to be able to query rrd data for alerts imo.

Back to reality.  Maybe a separate "device_stats" table would be good again, but not trying to store timeseries data like the old device_perf table.

eskyuu committed:
Local fix to allow me to roll bac schema

eskyuu commented:
@murrant I'm just getting back to this.  Are you ok with me moving the last_ping and last_ping_timetaken (rtt) value to the new table too?  This will obviously break any existing alert rules that people have, but it's needed if the goal is to avoid write locks.  Or would you prefer that I add both values to the new table, and then we mark the value in the device table as deprecated and remove it in a future release?

eskyuu committed:
Create a new device_stats table to store dynamic stats for devices.  Initial data is some RTT and packet loss stats/

eskyuu committed:
Formatting fix

eskyuu commented:
@murrant How does this look as a start for the device_stats table?  And do you have thoughts on the process to remove the existing fields from the devices table?

eskyuu committed:
Update DB schema

eskyuu committed:
Remove unneeded migration

eskyuu committed:
Remove refernce to device loss column

eskyuu committed:
Added some documentation around the averaging factor option

eskyuu committed:
Update to use LibrenmsConfig

murrant commented:
I think you need to rename the table. We had a table previously named device_stats. (out of caution)

eskyuu committed:
Add drop if exists on schema migration

eskyuu commented:
> I think you need to rename the table. We had a table previously named device_stats. (out of caution)

I couldn't think of a name that fits with other similar tables, but I've added a Schema::dropIfExists() before the create.  Does this work for you?

eskyuu committed:
Remove duplicate code

eskyuu committed:
Include the device ID when creating a new devicestats object

eskyuu committed:
Allow nulls in stats columns

murrant committed:
Refactor DeviceStats to extend DeviceRelatedModel

Updated DeviceStats to extend DeviceRelatedModel and removed the device relationship method.

murrant committed:
Remove unused BelongsTo import in DeviceStats

murrant committed:
Change ping RTT and loss fields to nullable

eskyuu committed:
Added the device_stats to the alert entities documentat

eskyuu committed:
Merge branch 'device-packetloss' of https://github.com/eskyuu/librenms into device-packetloss
